### DISCUSSION TITLE
How do I get Chrome to recognize new versions of imported js files?  (OJS)

### DISCUSSION THREAD

**Discussion by nwvhb** [quarto-dev/quarto-cli#6554] [Q&A] [answered]
### Description

If you import code from a javascript file and then make an update to that file, the cached version of the file will be used by Chrome on your hosted page.  Using quarto is there a way to specify the version of a javascript file?  For an ordinary html file you can specify the version in the script tag as shown. 

```html
<script type="text/javascript" src="code.js?1500"></script>`
```

In the Quarto code, I'm unsure of how to do the same. 

```qmd
```{ojs}
import  {x} from  './code.js'
```

Thank you,

**Comment by cscheid** [comment]
> If you import code from a javascript file and then make an update to that file, the cached version of the file will be used by Chrome on your hosted page.

I believe this is an issue with the caching settings on your server. Where are you experiencing that problem specifically?

**Reply by nwvhb** [reply]
I've hosted on Azure. The problem does not happen when run with local host.  When visiting the hosted site, it can be fixed manually by using CTRL+F5 in Chrome to replace the versions cached in your browser, but I'd rather not tell anyone to do that.  I think I can manually edit the generated html to fix it, but was wondering if there's a better way.

**Reply by cscheid** [reply]
This is a tricky bit of code. That specific javascript import is done [here](https://github.com/quarto-dev/quarto/blob/main/packages/ojs/quarto-ojs-runtime/src/ojs-connector.js#L223-L309).

I could attempt to allow `?query` query fragments in the import code, but this might not play nicely with the rest of observable's runtime.

I actually wonder if there's a better way to do this, by automatically doing an "integrity check" on those imports. We know those files at compile time, so we can compute their SHA256 hash. Then we use that as a query (which will do precisely the right thing wrt caching) and we can check for integrity on the way back.

It's going to take me a bit of time to implement this, but I think it's a sensible feature to consider.

**Reply by cscheid** [reply, marked as answer]
Tracking the issue here: https://github.com/quarto-dev/quarto-cli/issues/6556
